File: Introduction_to_Robotics_Course/Chapter15/Lesson1/Chapter15_Lesson1.cpp
#include <stdexcept>

struct SpeedSeparationConfig {
    double reaction_time;  // T_r
    double max_decel;      // a_max > 0
    double safety_margin;  // d_safe
};

double stoppingDistance(double v, const SpeedSeparationConfig& cfg) {
    if (cfg.max_decel <= 0.0) {
        throw std::runtime_error("max_decel must be positive");
    }
    if (v < 0.0) v = 0.0;
    return v * cfg.reaction_time + 0.5 * v * v / cfg.max_decel;
}

double safetyFunction(double d, double v, const SpeedSeparationConfig& cfg) {
    double d_stop = stoppingDistance(v, cfg);
    return d - d_stop - cfg.safety_margin;
}

bool hazardActive(double d, double v, const SpeedSeparationConfig& cfg) {
    return safetyFunction(d, v, cfg) < 0.0;
}

// Placeholder for integration with low-level robot controller:
void emergencyStop() {
    // Send stop command over fieldbus (e.g., CAN, EtherCAT) or digital output.
}

void monitorStep(double d, double v, const SpeedSeparationConfig& cfg) {
    if (hazardActive(d, v, cfg)) {
        emergencyStop();
    }
}
      

File: Introduction_to_Robotics_Course/Chapter15/Lesson1/Chapter15_Lesson1.py
from dataclasses import dataclass

@dataclass
class SpeedSeparationConfig:
    reaction_time: float   # T_r in seconds
    max_decel: float       # a_max > 0 (m/s^2)
    safety_margin: float   # d_safe (m)

def stopping_distance(v: float, cfg: SpeedSeparationConfig) -> float:
    """
    Compute stopping distance given relative speed v.
    v: relative speed (m/s), assumed v >= 0 when approaching.
    """
    if cfg.max_decel <= 0.0:
        raise ValueError("max_decel must be positive")
    return v * cfg.reaction_time + 0.5 * v * v / cfg.max_decel

def safety_function(d: float, v: float, cfg: SpeedSeparationConfig) -> float:
    """
    Safety function h(d, v) = d - d_stop(v) - d_safe.
    Safe if h >= 0, hazardous if h < 0.
    """
    d_stop = stopping_distance(max(v, 0.0), cfg)
    return d - d_stop - cfg.safety_margin

def hazard_active(d: float, v: float, cfg: SpeedSeparationConfig) -> bool:
    h = safety_function(d, v, cfg)
    return h < 0.0

def emergency_stop():
    # In a real robot, this would send a message to a safety PLC or low-level controller.
    print("!!! EMERGENCY STOP !!!")

def monitor_step(d: float, v: float, cfg: SpeedSeparationConfig):
    """
    One monitoring step; call this in a periodic loop.
    """
    if hazard_active(d, v, cfg):
        emergency_stop()
      
